### DISCUSSION TITLE
User Authentication (Laravel vs Custom)

### DISCUSSION THREAD

**Discussion by austintoddj** [austintoddj/canvas#959] [New Features]
I'd like to get feedback from those who use Canvas, and this seems like the best place to do it.

In the past, Canvas simply relied on Laravel's default authentication, and would utilize certain aspects of the standard User model. The pros of this were:
- No custom login/authentication was needed

The cons of this were:

- No user management (CRUD) via Canvas
- The user model in your app might not align with what Canvas is expecting (user_id vs id, etc)
- No built-in way to simply grab a user and their content, since the relationship isn't handled in the package

_[Recent issue](https://github.com/austintoddj/canvas/issues/939) with some related discussion_

I'd just like to know the general feel for what people would appreciate with this package, so give this:

👍 if you like it just the way it is
👎 if you'd like to see it reverted to the way it was

💡 Comment below if you can think of a better way to manage the User model.

**Comment by austintoddj** [comment]
Since it's looking like people prefer the former implementation of users in Canvas, my question then is how can you create a `CanvasUser` model that extends `app/Models/User` in the base app? I've seen Laravel Cashier do something similar with the [Billable trait](https://laravel.com/docs/8.x/billing#billable-model), not sure if we can utilize a similar solution.

At the end of the day, users just needs 3 things:
- CRUD functionality
- Role assignments
- Metadata (bio, summary, avatar, etc.)

Any thoughts or ideas here are greatly appreciated, I'm at a bit of a blocker on v7.0.0 until I figure this out.

**Reply by ReeceM** [reply]
Hi @austintoddj, I had implemented a way on an app that used the old method, but don't think there was CRUD from Canvas's side at that point? 

Also wasn't aware of the discussion here, I will have a look at what I had done on the app and share it here.

**Reply by frknasir** [reply]
> Since it's looking like people prefer the former implementation of users in Canvas, my question then is how can you create a `CanvasUser` model that extends `app/Models/User` in the base app? I've seen Laravel Cashier do something similar with the [Billable trait](https://laravel.com/docs/8.x/billing#billable-model), not sure if we can utilize a similar solution.
> 
> At the end of the day, users just needs 3 things:
> 
> * CRUD functionality
> * Role assignments
> * Metadata (bio, summary, avatar, etc.)
> 
> Any thoughts or ideas here are greatly appreciated, I'm at a bit of a blocker on v7.0.0 until I figure this out.

I think it's hard to tell people's preference from just 3 reactions. I, for one, am just seeing your post. and i absolutely like it the way it is. It became suitable for my project the moment the shift was made to custom login.

**Reply by olivsinz** [reply]
@Reecem

> Hi @austintoddj, I had implemented a way on an app that used the old method, but don't think there was CRUD from Canvas's side at that point?
> 
> Also wasn't aware of the discussion here, I will have a look at what I had done on the app and share it here.

Hi @ReeceM ! Please, how did you do it?

**Reply by ReeceM** [reply]
Hi @olivsinz, Are you referring to the current user database style or the old way? I was mainly referring to the old previous versions that used only the Laravel user

**Comment by austintoddj** [comment]
Closing this out with [Canvas v7.0.0](https://github.com/austintoddj/canvas/releases/tag/v7.0.0) — the design is settled.

- **Host app** owns authentication (login, password reset, sessions) via `CANVAS_GUARD` / your User model.
- **Canvas** owns access + author profile in `canvas_users` (role, bio, avatar, locale, digest, etc.).
- Optional `HasCanvasAccess` trait is host sugar only — not required for `/canvas` to work.
- Grant access with `canvas:make-admin` / `canvas:assign-role` (or the admin Users UI).

Full contract: [upgrade guide](https://github.com/austintoddj/canvas/blob/main/.github/UPGRADE.md).
